src/openMVG/multiview/solver_resection_p3p_nordberg.hpp

```diff
@@ -0,0 +1,248 @@
+      const Mat &X, // 3D points
+      std::vector<Mat34> *models);
+
+  // Compute the angular residual between the bearing vector and the 3d point projection vector
+  static double Error(
+      const Mat34 &P,
+      const Vec3 &bearing_vector,
+      const Vec3 &pt3D);
+};
+
+//-- Usable solver for robust estimation framework
+using PoseResectionKernel_P3P_Persson =
+    two_view::kernel::Kernel<
+        P3PSolver_Nordberg, // Model estimator
+        P3PSolver_Nordberg, // Error metric
+        Mat34>;
+
+static void gauss_newton_refineL(Vec3 &L,
+                          const double & a12, const double & a13, const double & a23,
+                          const double & b12, const double & b13, const double & b23)
+{
+
+  //TODO(RJ:) const expr makes it easier for the compiler to unroll
+  for (int i = 0; i < 5; ++i)
+  {
+    double l1 = L(0);
+    double l2 = L(1);
+    double l3 = L(2);
+    double r1 = l1 * l1 + l2 * l2 + b12 * l1 * l2 - a12;
+    double r2 = l1 * l1 + l3 * l3 + b13 * l1 * l3 - a13;
+    double r3 = l2 * l2 + l3 * l3 + b23 * l2 * l3 - a23;
+
+    if (std::abs(r1) + std::abs(r2) + std::abs(r3) < 1e-10)
+      break;
+
+    double dr1dl1 = 2.0 * l1 + b12 * l2;
+    double dr1dl2 = 2.0 * l2 + b12 * l1;
+
+    double dr2dl1 = 2.0 * l1 + b13 * l3;
+    double dr2dl3 = 2.0 * l3 + b13 * l1;
+
+    double dr3dl2 = 2.0 * l2 + b23 * l3;
+    double dr3dl3 = 2.0 * l3 + b23 * l2;
+
+    Vec3 r(r1, r2, r3);
+
+    // or skip the inverse and make it explicit...
+    {
+      double v0 = dr1dl1;
+      double v1 = dr1dl2;
+      double v3 = dr2dl1;
+      double v5 = dr2dl3;
+      double v7 = dr3dl2;
+      double v8 = dr3dl3;
+      double det = 1.0 / (-v0 * v5 * v7 - v1 * v3 * v8);
+
+      Mat3 Ji;
+      Ji << -v5 * v7, -v1 * v8, v1 * v5,
+            -v3 * v8, v0 * v8, -v0 * v5,
+             v3 * v7, -v0 * v7, -v1 * v3;
+      Vec3 L1 = Vec3(L) - det * (Ji * r);
+      //%l=l - g*H\G;%inv(H)*G
+      //L=L - g*J\r; 
+      //% works because the size is ok!
+      {
+        double l1 = L(0);
+        double l2 = L(1);
+        double l3 = L(2);
+        double r11 = l1 * l1 + l2 * l2 + b12 * l1 * l2 - a12;
+        double r12 = l1 * l1 + l3 * l3 + b13 * l1 * l3 - a13;
+        double r13 = l2 * l2 + l3 * l3 + b23 * l2 * l3 - a23;
+        if (std::abs(r11) + std::abs(r12) + std::abs(r13) > std::abs(r1) + std::abs(r2) + std::abs(r3))
+        {
+          break;
+        }
+        else
+          L = L1;
+      }
+    }
+  }
+};
+
+static inline bool root2real(const double & b, const double & c, double & r1, double & r2){
+    double v = b * b -4.0 * c;
+    if(v < 0.0){
+        r1 = r2 = 0.5 * b;
+        return false;
+    }
+    double y = std::sqrt(v);
+    if(b < 0.0){
+        r1 = 0.5 * (-b +y);
+        r2 = 0.5 * (-b -y);
+    }else{
+        r1= 2.0 * c / (-b + y);
+        r2= 2.0 * c / (-b - y);
+    }
+    return true;
+};
+
+static double cubick(const double & b, const double & c, const double & d){
```

Comment:
We already have a cubic solver, we could leverage the existing if we can https://github.com/openMVG/openMVG/blob/835f1e585377a55d4143548fa9181438f5d23889/src/openMVG/numeric/poly.h#L32

Reply:
I was not sure about that. Is there any risk to loose (or gain) accuracy or are the two solvers equivalents? (On my side, I'm not very concerned about speed as it's already very fast)

Reply:
It seems that both solver as a bit different after looking in details (1 vs 3 roots).
My recommendation is moving those couple of function to the cpp would be good to avoid having other interface using the functions.

Reply:
yes I reread the paper after 6 month, it seems to one of the advantage of the method over others p3p solvers.